app/src/main/java/com/pilot51/voicenotify/Service.kt

```diff
@@ -246,16 +247,30 @@ class Service : NotificationListenerService() {
 				info.ignoreReasons.add(IgnoreReason.EMPTY_MSG)
 			}
 			if (ttsMsg != null) {
+				fun containsOrMatchesRegex(it: String): Boolean {
```

Comment:
I'm not really a fan of creating local functions like this. I think moving it to the companion object as an extension would be better:
```
private fun String.containsOrMatchesRegex(it: String): Boolean {
	return if (it.startsWith(Constants.REGEX_PREFIX)) {
		try {
			Regex(it.removePrefix(Constants.REGEX_PREFIX)).containsMatchIn(this)
		} catch (e: PatternSyntaxException) {
			e.printStackTrace()
			CoroutineScope(Dispatchers.Main).launch {
				Toast.makeText(appContext, R.string.malformed_regex, Toast.LENGTH_LONG).show()
			}
			false
		}
	} else contains(it, true)
}
```

Usage: `ttsMsg.containsOrMatchesRegex(it)`

Reply:
Well, local functions are there for a reason - they are releveant only to the code using them. We could just duplicate the code and ditch the function, but I think it is less elegant.

Also, the toast should be only when saving the regex no? There is not reason to put it on the service

Reply:
```
I agree, case sensitive allows for more flexibility and I'm sure anyone who wants to use regex would appreciate it.
Unfortunately, it would appear you forgot to update the code in both places when you changed the strings to say case sensitive.
```
Sorry, but I didn't understand what I forgot, could you please refer me to the file?

Reply:
I believe that everything has it's time and place, and sometimes nested functions might be the best solution, like in this case.
I will change it to lambda.
You are welcome to refactor if you would like, I have no problem with that. I just think that a code refactor should be backed by a good reason, and I don't really see the benefit here.

I will try to move it on save, but I am a bit short on time so if I get stuck I will use the lazy solution.

I saw, thanks

Reply:
It's just a matter of familiarity. I don't think I've seen anyone use a local function before even though I work with other experienced Kotlin devs and there's plenty of opportunity to do so in those much larger projects, so it looks weird to me. Maybe that's because we all come from a Java background and it's almost never mentioned in Kotlin documentation and examples. I haven't given it much thought until now, so I'm looking around for other perspectives from the larger Kotlin community and I may change my mind. Feel free to leave it a local function if you prefer.

On the other hand, Kotlin allows functions and properties at the file level outside of any class, again unlike Java. It felt wrong at first, but thanks to how common it is for Compose, I've gotten used to it.

Reply:
I actually come from a c# background, I don't really know Kotlin 😄

I rarely use local functions aswell, but on some occasions they are helpful.